cow_amm/uni_v2_amm_lp_and_tvl_4047194.sql

```diff
@@ -0,0 +1,163 @@
+-- Computes the TVL, lp token total supply and lp token price of a uniswap pool over time
+-- Parameters
+--  {{token_a}} - either token of the desired uni pool
+--  {{token_b}} - other token of the desired uni pool
+--  {{start}} - date as of which the analysis should run
+
+-- Given that we might not have records every day in the source data (e.g. not every day the lp supply may change), 
+-- but still want to visualize development on a per day basis,  we create an auxiliary table with one record per 
+-- day between `start` and `now`
+with date_range as (
+    select t.day
+    from
+        unnest(sequence(
+            date(timestamp '{{start}}'),
+            date(now())
+        )) t (day) --noqa: AL01
+),
+
+-- Finds the uniswap v2 pool address given tokens specified in query parameters (regardless of order)
+pool as (
+    select
+        pool as contract_address,
+        token0,
+        token1
+    from uniswap_ethereum.pools
+    where
+        ((token0 = {{token_a}} and token1 = {{token_b}}) or (token1 = {{token_a}} and token0 = {{token_b}}))
+        and version = 'v2'
+    limit 1
+),
+
+-- per day lp token total supply changes of the uniswap pool by looking at burn/mint events
+lp_supply_delta as (
+    select
+        date(evt_block_time) as "day",
+        sum(case when "from" = 0x0000000000000000000000000000000000000000 then value else -value end) as delta
+    from erc20_ethereum.evt_transfer
+    where
+        contract_address = (select contract_address from pool)
+        and ("from" = 0x0000000000000000000000000000000000000000 or "to" = 0x0000000000000000000000000000000000000000)
+    group by 1
+),
+
+-- per day lp token total supply by summing up all deltas (may have missing records for some days)
+lp_total_supply_incomplete as (
+    select
+        day,
+        sum(delta) over (order by day) as total_lp
+    from lp_supply_delta
+),
+
+-- lp token total supply without date gaps
+lp_total_supply as (
+    select
+        day,
+        total_lp
+    from (
+        -- join full date range with potentially incomplete data. This results in many rows per day (all total supplies on or before that day)
+        -- rank() is then used to order join candidates by recency (rank = 1 is the latest lp supply)
+        select
+            date_range.day,
+            total_lp,
+            rank() over (partition by (date_range.day) order by lp.day desc) as latest
+        from date_range
+        inner join lp_total_supply_incomplete as lp
+            on date_range.day >= lp.day
+            -- performance optimisation: this assumes one week prior to start there was at least one lp supply change event
+            and lp.day >= (timestamp '{{start}}' - interval '7' day)
+    )
+    where latest = 1
```

Comment:
> is it supposed to take the previous available value if the data is missing for a certain date?

Yes exactly. Can you elaborate why you think this may not be working? 

Wouldn't we in your suggestion still miss a record if there are missing values for two days in a row (which is the case for less frequently traded pairs)? Cf minimal example query: https://dune.com/queries/4055768

Reply:
Previous approach did not work for me initially, but after rerunning it again it seems to be fine.
Initially, if I selected the date with the missing values, it would not find any corresponding previous values, but I cannot replicate it anymore, so all fine from my size now. Adjusted my query as well.